copulas/__init__.py

```diff
@@ -21,19 +21,32 @@ class NotFittedError(Exception):
 
 
 @contextlib.contextmanager
-def random_seed(seed):
-    """Context manager for managing the random seed.
+def set_random_state(random_state, set_model_random_state):
+    """Context manager for managing the random state.
 
     Args:
-        seed (int):
-            The random seed.
+        random_state (int or np.random.RandomState):
+            The random seed or RandomState.
+        set_model_random_state (function):
+            Function to set the random state on the model.
     """
-    state = np.random.get_state()
-    np.random.seed(seed)
+    original_state = np.random.get_state()
+
+    if isinstance(random_state, int):
+        desired_state = np.random.RandomState(seed=random_state).get_state()
+    elif isinstance(random_state, np.random.RandomState):
+        desired_state = random_state.get_state()
+    elif not isinstance(random_state, tuple):
+        raise TypeError(f'RandomState {random_state} is an unexpected type. '
+                        'Expected to be int, np.random.RandomState, or tuple.')
+
+    np.random.set_state(desired_state)
```

Comment:
Bug: If `random_state` is a `tuple`, `desired_state` is never assigned any value. I think that it would be simpler to just re-use the `random_state` variable name instead of `desired_state`